src/main/java/com/limechain/babe/BlockProductionVerifier.java

```diff
@@ -39,19 +55,45 @@ public boolean verifyAuthorship(BlockHeader blockHeader, EpochData currentEpochD
         int authorityIndex = babePreDigest.map(BabePreDigest::getAuthorityIndex).map(Math::toIntExact).get();
         Authority verifyingAuthority = authorities.get(authorityIndex);
 
-        VrfOutputAndProof vrfOutputAndProof = VrfOutputAndProof.wrap(
-                babePreDigest.get().getVrfOutput(), babePreDigest.get().getVrfProof());
-        VerifySignature signature = new VerifySignature(
-                signatureData, blockHeader.getHashBytes(), verifyingAuthority.getPublicKey(), Key.SR25519);
-
-        return Sr25519Utils.verifySignature(signature) &&
-                verifySlotWinner(authorityIndex,
-                        authorities,
-                        currentEpochIndex,
-                        randomness,
-                        descriptor,
-                        currentSlotNumber,
-                        vrfOutputAndProof);
+        if (!checkBlockEquivocation(verifyingAuthority, blockHeader, runtime, currentSlotNumber)) {
+            return false;
+        }
+
+        VrfOutputAndProof vrfOutputAndProof = VrfOutputAndProof.wrap(babePreDigest.get().getVrfOutput(), babePreDigest.get().getVrfProof());
+        VerifySignature signature = new VerifySignature(signatureData, blockHeader.getHashBytes(), verifyingAuthority.getPublicKey(), Key.SR25519);
+
+        boolean isVerified = Sr25519Utils.verifySignature(signature) &&
+                verifySlotWinner(authorityIndex, authorities, currentEpochIndex, randomness, descriptor, currentSlotNumber, vrfOutputAndProof);
+
+        if (isVerified) {
+            currentSlotAuthorBlockMap.put(StringUtils.toHexWithPrefix(verifyingAuthority.getPublicKey()),blockHeader);
+        }
+
+        return isVerified;
+    }
+
+    private boolean checkBlockEquivocation(Authority authority,
+                                           BlockHeader blockHeader,
+                                           Runtime runtime,
+                                           BigInteger currentSlotNumber) {
+        String hexPublicKey = HexUtils.toHexString(authority.getPublicKey());
+        if (currentSlotAuthorBlockMap.containsKey(hexPublicKey)) {
+            BlockHeader firstBlockHeader = currentSlotAuthorBlockMap.get(hexPublicKey);
+            if (!firstBlockHeader.equals(blockHeader)) {
```

Comment:
Please, override the equals() method for BlockHeader, because currently it uses Object equals() method, which return true if we are comparing the one object instance to itself.
![Screenshot 2024-12-05 at 0 02 48](https://github.com/user-attachments/assets/dbbe8113-3f91-4ebb-9722-947db8c2ea03)

Reply:
I just realized that the ```firstBlockHeader.equals(blockHeader)``` check might not be necessary after all. You could directly use ```firstBlock.getHash().equals(blockHeader.getHash())``` instead. Please feel free to choose the approach you think works best.